functions/go/enable-gcp-services/gcpservices/project_service_list.go

```diff
@@ -0,0 +1,230 @@
+package gcpservices
+
+import (
+	"fmt"
+	"path"
+	"strconv"
+
+	"sigs.k8s.io/kustomize/kyaml/kio/filters"
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const (
+	projectServiceListKind       = "ProjectServiceList"
+	projectServiceListAPIVersion = "blueprints.cloud.google.com/v1alpha1"
+	serviceUsageAPIVersion       = "serviceusage.cnrm.cloud.google.com/v1beta1"
+	serviceUsageKind             = "Service"
+	managedByAnnotation          = "blueprints.cloud.google.com/managed-by-enable-gcp-services"
+)
+
+// ignoreAnnotations is used to ignore any annotation that should not be added to generated Service CRs
+var ignoreAnnotations = map[string]bool{kioutil.PathAnnotation: true, filters.LocalConfigAnnotation: true}
+
+// ProjectServiceListRunner discovers ProjectServiceList CRs to generate Service CRs
+type ProjectServiceListRunner struct {
+	results []Result
+}
+
+// ProjectServiceList is the functionConfig for generating services
+type ProjectServiceList struct {
+	yaml.ResourceMeta `json:",inline" yaml:",inline"`
+	Spec              struct {
+		Services  []string `json:"services" yaml:"services"`
+		ProjectID string   `json:"projectID,omitempty" yaml:"projectID,omitempty"`
+	} `json:"spec" yaml:"spec"`
+	results []Result
+}
+
+// Result represents an action performed by ProjectServiceList on a resource
+type Result struct {
+	ResourceRef yaml.ResourceIdentifier
+	FilePath    string
+	Action      string
```

Comment:
It seems we have a limited number of `Action`s.
We can do something like `type ActionType string` and then define `GeneratedService` and `PruneService` action types.

Reply:
done